File: DataEngineering/07_Data_Transformation/projects/golang/01_sql_templating_engine/main.go
package main

import (
	"bytes"
	"context"
	"database/sql"
	"fmt"
	"log"
	"os"
	"strings"
	"text/template"
	"time"

	_ "github.com/lib/pq"
)

type QueryContext struct {
	Columns    []string
	Table      string
	Conditions []string
}

// GenerateSQL parses the template and data to produce a SQL string.
func GenerateSQL(tmplStr string, data QueryContext) (string, error) {
	tmpl, err := template.New("sql").Funcs(template.FuncMap{
		"join": strings.Join,
	}).Parse(tmplStr)
	if err != nil {
		return "", fmt.Errorf("failed to parse template: %w", err)
	}

	var buf bytes.Buffer
	if err := tmpl.Execute(&buf, data); err != nil {
		return "", fmt.Errorf("failed to execute template: %w", err)
	}
	return buf.String(), nil
}

// ExecuteQuery runs the query on the provided DB connection.
func ExecuteQuery(ctx context.Context, db *sql.DB, query string) error {
	log.Printf("Executing query: \n%s", strings.TrimSpace(query))
	
	rows, err := db.QueryContext(ctx, query)
	if err != nil {
		return fmt.Errorf("query execution failed: %w", err)
	}
	defer rows.Close()

	var count int
	for rows.Next() {
		count++
	}

	if err := rows.Err(); err != nil {
		return fmt.Errorf("row iteration error: %w", err)
	}

	log.Printf("Query returned %d rows", count)
	return nil
}

func main() {
	dsn := os.Getenv("DATABASE_URL")
	if dsn == "" {
		dsn = "postgres://user:password@localhost:5432/analytics_db?sslmode=disable"
	}

	db, err := sql.Open("postgres", dsn)
	if err != nil {
		log.Fatalf("Failed to open DB connection: %v", err)
	}
	defer db.Close()

	db.SetMaxOpenConns(10)
	db.SetMaxIdleConns(5)
	db.SetConnMaxLifetime(time.Minute * 5)

	sqlTmpl := `SELECT
{{join .Columns ", "}}
FROM {{.Table}}
WHERE {{range $i, $cond := .Conditions}}{{if $i}} AND {{end}}{{$cond}}{{end}};`

	data := QueryContext{
		Columns:    []string{"id", "name", "amount", "transaction_date"},
		Table:      "raw_transactions",
		Conditions: []string{"amount > 100", "status = 'completed'"},
	}

	query, err := GenerateSQL(sqlTmpl, data)
	if err != nil {
		log.Fatalf("Template generation error: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	if err := ExecuteQuery(ctx, db, query); err != nil {
		log.Printf("Note: Database not reachable, skipping execution. Query was: \n%s", query)
	}
}

File: DataEngineering/07_Data_Transformation/projects/python/01_sql_templating_engine/main.py
import os
import logging
from typing import List, Dict, Any
from jinja2 import Environment, FileSystemLoader, Template
from sqlalchemy import create_engine, text
from sqlalchemy.engine import Engine
from sqlalchemy.exc import SQLAlchemyError

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")
logger = logging.getLogger(__name__)

class SQLTemplatingEngine:
    """
    A production-grade SQL templating engine that dynamically generates 
    and executes SQL queries using Jinja2 and SQLAlchemy.
    """
    def __init__(self, db_url: str):
        self.engine: Engine = create_engine(db_url, pool_pre_ping=True, pool_size=5)
        
    def generate_query(self, template_str: str, context: Dict[str, Any]) -> str:
        """Generates a SQL query from a Jinja2 template string."""
        template = Template(template_str)
        return template.render(**context)

    def execute_query(self, query: str) -> List[Dict[str, Any]]:
        """Executes the generated SQL query and returns results as a list of dictionaries."""
        try:
            with self.engine.connect() as connection:
                logger.info(f"Executing Query: {query.strip()}")
                result = connection.execute(text(query))
                
                if result.returns_rows:
                    return [dict(row._mapping) for row in result]
                return []
        except SQLAlchemyError as e:
            logger.error(f"Database execution failed: {e}")
            raise

if __name__ == "__main__":
    db_url = os.getenv("DATABASE_URL", "postgresql://user:password@localhost:5432/analytics_db")
    sql_template = """
    SELECT
        {{ columns | join(', ') }}
    FROM {{ table_name }}
    WHERE {% for condition in conditions %}
        {{ condition }}{% if not loop.last %} AND {% endif %}
    {% endfor %}
    """
    
    context = {
        "columns": ["id", "name", "amount", "transaction_date"],
        "table_name": "raw_transactions",
        "conditions": ["amount > 100", "status = 'completed'"]
    }
    
    templater = SQLTemplatingEngine(db_url)
    try:
        query = templater.generate_query(sql_template, context)
        # Using a try/except for execution to allow it to run without a real DB
        try:
            results = templater.execute_query(query)
            logger.info(f"Retrieved {len(results)} rows.")
        except SQLAlchemyError:
            logger.warning("Could not execute query. Ensure a real database is running.")
    except Exception as e:
        logger.error("Failed to run templated query pipeline.", exc_info=True)
